mail_tracking/models/res_partner.py

```diff
@@ -40,10 +40,15 @@ def email_bounced_set(self, tracking_emails, reason, event=None):
         return partners.write({'email_bounced': True})
 
     def write(self, vals):
-        email = vals.get('email')
-        if email is not None:
-            vals['email'] = email.lower() if email else False
-            vals['email_bounced'] = (
-                bool(email) and
-                self.env['mail.tracking.email'].email_is_bounced(email))
+        if 'email' not in vals:
+            return super(ResPartner, self).write(vals)
+        vals['email'] = vals['email'].lower() if vals['email'] else False
+        mte_obj = self.env['mail.tracking.email']
+        if not mte_obj.email_is_bounced(vals['email']):
+            vals['email_bounced'] = False
+            return super(ResPartner, self).write(vals)
+        res = mte_obj._email_last_tracking_state(vals['email'])
+        tracking = mte_obj.browse(res.get('id'))
+        event = tracking.event_ids[:1]
+        self.email_bounced_set(reason=event.reason, event=event)
```

Comment:
Why logging this on write?

Reply:
Well, we're checking that the email we're setting to the partner isn't already in a bounce state and logging that state to the partner. For instance, you merge a partner into another, or you use an email that is in hard-bounce state. This will only be checked when the email is set.